### DISCUSSION TITLE
contains_eager with aliased table raises "SAWarning: SELECT statement has a cartesian product between FROM element(s)"

### DISCUSSION THREAD

**Discussion by Feijo** [sqlalchemy/sqlalchemy#7681] [Usage Questions] [answered]
> SQLAlchemy==1.4.31

I have 3 tables: **Campaign**, **Product**, **ProductPicture**.

A campaign contains multiple products. A product contains multiple pictures, however, pictures have a date range, and there's only one "current" picture at a time. I'm trying to load a list of campaigns, with products, and just the current picture.

Full running example

```python
from datetime import date
from itertools import product
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import Integer, String, Date
from sqlalchemy import Table
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import aliased
from sqlalchemy.orm import contains_eager
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()

metadata = Base.metadata

campaign_table = Table(
    "campaign",
    metadata,
    Column("id", Integer, primary_key=True),
)

product_table = Table(
    "product",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("campaign_id", ForeignKey("campaign.id")),
)

product_picture = Table(
    "picture",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("product_id", ForeignKey("product.id")),
    Column("start_date", Date),
    Column("end_date", Date),
)

class Campaign(Base):
    __table__ = campaign_table
    id: int
    products = relationship("Product")

class Product(Base):
    __table__ = product_table
    id: int
    current_picture = relationship("ProductPicture", uselist=False, viewonly=True)


class ProductPicture(Base):
    __table__ = product_picture
    id: int
    product_id: int
    start_date: date
    end_date: date


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(Campaign(products=[Product(), Product()]))
s.commit()
s.close()


def current_picture_query(s):
    today = date.today()

    columns = [c for c in product_picture.columns if c.name != 'end_date']

    current_picture = (s.query(*columns, func.max(ProductPicture.start_date).label('product_picture_end_date'))
                        .filter(ProductPicture.start_date <= today)
                        .group_by(ProductPicture.id, ProductPicture.start_date)
                        .subquery(with_labels=True)
                        .alias('current_picture'))

    return current_picture


current_picture_alias = current_picture_query(s)

query = (
    s.query(Campaign)
    .outerjoin(Campaign.products)
    .outerjoin(current_picture_alias, current_picture_alias.c.picture_product_id == Product.id)
    .options(
        contains_eager(Campaign.products)
        # THE PROBLEM SEEMS TO BE IN THIS SECOND CONTAINS_EAGER
        .options(contains_eager(Product.current_picture, alias=current_picture_alias))
    )
)

records = query.all()
s.close()

```

> SAWarning: SELECT statement has a cartesian product between FROM element(s) "picture" and FROM element "product".  Apply join condition(s) between each element to resolve.
>   records = query.all()

**Comment by zzzeek** [comment, marked as answer]
this looked like so many other contains_eager() scenarios that are sometimes not solvable, but then i looked more and it is.  cartesian product is when it's sticking the mapped table and not your subquery into the query:

```
SELECT current_picture.picture_id AS current_picture_picture_id, current_picture.picture_product_id AS current_picture_picture_product_id, current_picture.picture_start_date AS current_picture_picture_start_date, picture.end_date AS picture_end_date, product.id AS product_id, product.campaign_id AS product_campaign_id, campaign.id AS campaign_id 
FROM picture, campaign LEFT OUTER JOIN product ON campaign.id = product.campaign_id LEFT OUTER JOIN (SELECT picture.id AS picture_id, picture.product_id AS picture_product_id, picture.start_date AS picture_start_date, max(picture.start_date) AS product_picture_end_date 
FROM picture 
WHERE picture.start_date <= ? GROUP BY picture.id, picture.start_date) AS current_picture ON current_picture.picture_product_id = product.id
```

this is the giveaway:

```
FROM picture, campaign LEFT OUTER JOIN ...
```

contains_eager will do that when it is asking for columns for the entity it's being asked to load.  we can see what it's asking for from "picture":

```
picture.end_date AS picture_end_date
```

which makes sense because it's being omitted:

```
    columns = [c for c in product_picture.columns if c.name != 'end_date']
```

tell the ORM to not ask for this column using defer():

```py
    .options(
        contains_eager(Campaign.products)
        .options(contains_eager(Product.current_picture, alias=current_picture_alias).defer(ProductPicture.end_date))
    )
```


(yes I am sort of amazed that works)

**Reply by Feijo** [reply]
Oh wow, now that you pointed that out, it makes perfect sense why the query needed to go to the `picture` table again, because it didn't have the `end_date` field in the subquery. 

Indeed, the warning is gone but I think there's a small problem still. My idea with this code:
`s.query(*columns, func.max(ProductPicture.start_date).label('product_picture_end_date')`

Was to use this calculated value for the `end_date`, instead of the `end_date` that is in the row.

I think that now we are using `defer`, it will just `defer` the loading of the field, but it will lazy load the `end_date` from the row, not the calculated value I have (`func.max`).

Edit: I just realised my `func.max` logic is a bit wrong, but it doesn't change the expected outcome.

**Reply by zzzeek** [reply]
This is really something that could be fixed in contains_eager().   we gave it an alias object to look for columns within.  it should be doable that it doesnt try to pull columns from other tables.  I've never thought to implement this since this is usually just the joinedload() logic where this kind of thing doesnt happen.    not sure how often this use case is.  at least the cartesian warning gives the user a heads up.

**Reply by zzzeek** [reply]
OK you want to use that "end_date" label, try loading from an  aliased(ProductPicture, my_subquery, adapt_on_names=True)   that should look for "end_date" by name.   edge case stuff tho

**Reply by Feijo** [reply]
Thanks @zzzeek, I agree it might be an edge case. I'll play around with your suggestions. Thanks for the help!
